patches/api/0390-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,476 @@
++            this.registerRedirect(dispatcherRoot, root, identifier + ":" + alias);
++        }
++    }
++
++    private void registerRedirect(RootCommandNode<CommandSourceStack> dispatcherRoot, PluginCommandNode node, String name) {
++        PluginCommandNode aliasRoot = CommandBuilder.of(this.plugin, name)
++            .description(this.description)
++            .redirect(node)
++            .build();
++
++        dispatcherRoot.addChild(aliasRoot);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..3d2ca2d0e092c1aa15adf7a32da8567c614aa43f
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+@@ -0,0 +1,23 @@
++package io.papermc.paper.command.brigadier;
++
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.command.CommandSender;
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Nullable;
++
++@ApiStatus.NonExtendable
++public interface CommandSourceStack {
++
++    @Nullable
++    Entity getBukkitEntity();
++
++    @Nullable
++    World getBukkitWorld();
++
++    @Nullable
++    Location getBukkitLocation();
++
++    CommandSender getBukkitSender();
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..839927c90a095674a332a8f61e3c279c4c08129a
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializer.java
+@@ -0,0 +1,18 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.Message;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.serializer.ComponentSerializer;
++
++public interface MessageComponentSerializer extends ComponentSerializer<Component, Component, Message> {
++
++    /**
++     * A component serializer that converts it to a brigadier compatible type.
++     *
++     * @return serializer instance
++     */
++    static MessageComponentSerializer message() {
++        return MessageComponentSerializerHolder.PROVIDER;
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..61ee6e72643dbd1e46dce362adddc9936631df95
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/MessageComponentSerializerHolder.java
+@@ -0,0 +1,10 @@
++package io.papermc.paper.command.brigadier;
++
++import java.util.ServiceLoader;
++
++class MessageComponentSerializerHolder {
```

Comment:
This can just be an inner class (`Holder`) of MessageComponentSerializer, with a non public static final field inside it.